src/devices/Board/Board.cs

```diff
@@ -0,0 +1,708 @@
+
+using System;
+using System.Collections.Generic;
+using System.Device.Gpio;
+using System.Device.Gpio.Drivers;
+using System.Device.I2c;
+using System.Device.Pwm;
+using System.Device.Spi;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Microsoft.Win32;
+
+namespace Iot.Device.Board
+{
+    /// <summary>
+    /// Base class for all board abstractions.
+    /// A "board" is a piece of hardware that offers low-level interfaces to other devices. Typically, it has GPIO pins and one or multiple SPI or I2C busses.
+    /// There should be exactly one instance of a board class per hardware component in an application, but it is possible to work with multiple boards
+    /// at once (i.e. when having a GPIO expander connected to the Raspberry Pi)
+    /// </summary>
+    public abstract class Board : MarshalByRefObject, IDisposable
+    {
+        // See comment at GetBestDriverForBoardOnWindows. This should get some specific factory pattern
+        private const string BaseBoardProductRegistryValue = @"SYSTEM\HardwareConfig\Current\BaseBoardProduct";
+        private const string RaspberryPi2Product = "Raspberry Pi 2";
+        private const string RaspberryPi3Product = "Raspberry Pi 3";
+        private const string HummingBoardProduct = "HummingBoard-Edge";
+
+        private readonly PinNumberingScheme _defaultNumberingScheme;
+        private readonly object _pinReservationsLock;
+        private readonly Dictionary<int, List<PinReservation>> _pinReservations;
+        private readonly Dictionary<int, I2cBusManager> _i2cBuses;
+        private bool _initialized;
+        private bool _disposed;
+
+        /// <summary>
+        /// Constructs a board instance with the given default numbering scheme.
+        /// All methods will use the given numbering scheme, unless another scheme is explicitly given in a call.
+        /// </summary>
+        /// <param name="defaultNumberingScheme">Default numbering scheme for the board.</param>
+        /// <remarks>
+        /// The constructor will never throw an exception. Call <see cref="Initialize"/> to initialize the hardware and check
+        /// whether an instance can actually run on the current hardware.
+        /// </remarks>
+        protected Board(PinNumberingScheme defaultNumberingScheme)
+        {
+            _defaultNumberingScheme = defaultNumberingScheme;
+            _pinReservations = new Dictionary<int, List<PinReservation>>();
+            _i2cBuses = new Dictionary<int, I2cBusManager>();
+            _pinReservationsLock = new object();
+            _initialized = false;
+            _disposed = false;
+        }
+
+        /// <summary>
+        /// (Temporary) Register to this event to receive log messages.
+        /// </summary>
+        public event Action<string, Exception?>? LogMessages;
+
+        /// <summary>
+        /// True if the board instance is initialized
+        /// </summary>
+        protected bool Initialized => _initialized;
```

Comment:
this feels like it shouldn't be public

Reply:
This is protected, because derived classes should also be able to throw if this is false.